src/engine/enginebuffer.cpp

```diff
@@ -734,15 +735,27 @@ void EngineBuffer::process(CSAMPLE* pOutput, const int iBufferSize) {
 
     bool bCurBufferPaused = false;
     double rate = 0;
+    double sample_rate = m_pSampleRate->get();
 
     bool bTrackLoading = load_atomic(m_iTrackLoading) != 0;
     if (!bTrackLoading && m_pause.tryLock()) {
         ScopedTimer t("EngineBuffer::process_pauselock");
-        float sr = m_pSampleRate->get();
 
         double baserate = 0.0;
-        if (sr > 0) {
-            baserate = ((double)m_file_srate_old / sr);
+        if (sample_rate > 0) {
+            baserate = ((double)m_file_srate_old / sample_rate);
+        }
+
+        // If the sample rate has changed, force Rubberband to reset so that
+        // it doesn't reallocate when the user engages keylock during playback.
+        // We do this even if rubberband is not active.  If it is active, the
+        // regular call to setScaleParameters will trigger the reset.
+        if (sample_rate != m_dPrevSampleRate) {
+            if (m_pScaleRB != NULL && m_pScale != m_pScaleRB) {
+                double tempo = 1.0;
+                double pitch = 1.0;
+                m_pScaleRB->setScaleParameters(sample_rate, 1.0, &tempo, &pitch);
+            }
```

Comment:
setScaleParameters() has already a sample rate change set and does a lot more than just initialize Rubberband
```
if (m_iSampleRate != iSampleRate) {
initializeRubberBand(iSampleRate);
m_iSampleRate = iSampleRate;
}
```

I would prefer to introduce an extra function that does the buffer hack only. 

What about removing the sample rate parameter from the  setScaleParameters call and introduce an extra function for this.
All scalers will benefit from it.